pydbtools/utils.py

```diff
@@ -1,18 +1,144 @@
+from typing import Tuple
+
 import numpy as np
 
 from gluejobutils.s3 import s3_path_to_bucket_key, check_for_s3_file
 import os
+import re
+import sqlparse
 from s3fs import S3FileSystem
 
+import boto3
+from botocore.credentials import InstanceMetadataProvider, InstanceMetadataFetcher
+
 # pydbtools will create a new a new S3 object (then delete it post read). In the first call read
 # the cache is empty but then filled. If pydbtools is called again the cache is referenced and
 # you get an NoFileError.
 # Setting cachable to false fixes this. cachable is class object from fsspec.AbstractFileSystem
 # which S3FileSystem inherits.
 S3FileSystem.cachable = False
 
+# Get role specific path for athena output
+bucket = "mojap-athena-query-dump"
+
+temp_database_name_prefix = "mojap_de_temp_"
+
+
+def check_temp_query(sql_query: str):
+    """
+    Checks if a query to a temporary table
+    has had __temp__ wrapped in quote marks.
+
+    Args:
+        sql_query (str): an SQL query
+
+    Raises:
+        ValueError
+    """
+    if re.findall(r'["|\']__temp__["|\']\.', sql_query.lower()):
+        raise ValueError(
+            "When querying a temporary database, __temp__ should not be wrapped in quotes"
+        )
+
+
+def clean_query(sql_query: str) -> str:
+    """
+    removes trailing whitespace, newlines and final
+    semicolon from sql_query for use with
+    sqlparse package
+
+    Args:
+        sql_query (str): The raw SQL query
+
+    Returns:
+        str: The cleaned SQL query
+    """
+    return " ".join(sql_query.splitlines()).strip().rstrip(";")
+
+
+def replace_temp_database_name_reference(sql_query: str, database_name: str) -> str:
+    """
+    Replaces references to to the users temp database __temp__
+    with the database_name string provided.
+
+    Args:
+        sql_query (str): The raw SQL query as a string
+        database_name (str): The database name to replace __temp__
+
+    Returns:
+        str: The new SQL query which is sent to Athena
+    """
+    # check query is valid and clean
+    parsed = sqlparse.parse(clean_query(sql_query))
+    new_query = []
+    for query in parsed:
+        print(query)
+        check_temp_query(str(query))
+        for word in str(query).strip().split(" "):
+            if "__temp__." in word.lower():
+                word = word.lower().replace("__temp__.", f"{database_name}.")
+            new_query.append(word)
+        if ";" not in new_query[-1]:
+            last_entry = new_query[-1] + ";"
+        else:
+            last_entry = new_query[-1]
+        del new_query[-1]
+        new_query.append(last_entry)
+    return " ".join(new_query)
```

Comment:
Could 81-87 be simpler, given that clean_query should already have stripped semicolons from the end? I guess double-checking doesn't hurt.

Reply:
Yeah cleaning up queries was a real pain! Will take a look to see if it can be tidier here